src/Common/SimulationIC.hpp

```diff
@@ -82,106 +83,98 @@ void Simulation<ndim>::GenerateIC(void)
   }
 
 
-  // If not a restart, generate initial conditions either from external file or created on the fly.
+  // Gnerate initial conditions either from external file
   //-----------------------------------------------------------------------------------------------
-  Ic<ndim> icGenerator(this, hydro, invndim);
-
   if (ic == "file") {
     ReadSnapshotFile(simparams->stringparams["in_file"], simparams->stringparams["in_file_form"]);
     rescale_particle_data = true;
     this->initial_h_provided = false;
+    icGenerator = new NullIc<ndim>(this, hydro, invndim);
   }
+  // Generate initial conditions via python scripts
   //-----------------------------------------------------------------------------------------------
-  else if (ic == "bb") {
-    icGenerator.BossBodenheimer();
-  }
-  else if (ic == "binary") {
-    icGenerator.BinaryStar();
+  else if (ic == "python") {
+    icGenerator = new NullIc<ndim>(this, hydro, invndim);
+    return;
   }
+  // Generate initial conditions on-the-fly via various IC class functions
+  //-----------------------------------------------------------------------------------------------
   else if (ic == "binaryacc") {
-    icGenerator.BinaryAccretion();
-  }
-  else if (ic == "blastwave") {
-    icGenerator.BlastWave();
-  }
-  else if (ic == "blob") {
-    icGenerator.BlobTest();
+    icGenerator = new BinaryAccretionIc<ndim>(this, hydro, invndim);
   }
   else if (ic == "bondi") {
-    icGenerator.BondiAccretion();
+    icGenerator = new BondiAccretionIc<ndim>(this, hydro, invndim);
   }
-  else if (ic == "box") {
-    icGenerator.UniformBox();
+  else if (ic == "bb" || ic == "bossbodenheimer") {
+    icGenerator = new BossBodenheimerIc<ndim>(this, hydro, invndim);
   }
   else if (ic == "cdiscontinuity") {
-    icGenerator.ContactDiscontinuity();
+    icGenerator = new ContactDiscontinuityIc<ndim>(this, hydro, invndim);
+  }
+  else if (ic == "dustybox") {
+    icGenerator = new DustyBoxIc<ndim>(this, hydro, invndim);
+  }
+  else if (ic == "evrard") {
+    icGenerator = new EvrardCollapseIc<ndim>(this, hydro, invndim);
   }
-  else if (ic == "ewaldsine" || ic == "jeans" ||
-           ic == "ewaldsine2" || ic == "ewaldslab" || ic == "ewaldcylinder") {
-    icGenerator.EwaldDensity();
+  else if (ic == "ewaldsine" || ic == "ewaldsine2" ||
+           ic == "ewaldslab" ||  ic == "ewaldcylinder" ||
+           ic == "jeans") {
+    icGenerator = new EwaldIc<ndim>(this, hydro, invndim);
   }
-  else if (ic == "gaussianring") {
-	  icGenerator.GaussianRing();
```

Comment:
Actually, that diff you've highlighted shows I'm talking rubbish since it was there and was deleted!  Hmmmm, maybe I deleted them accidentally and didn't notice.  Anyway, I'll put it back in there now!